TestingLowerBounds/Testing/Risk.lean

```diff
@@ -33,92 +33,88 @@ namespace ProbabilityTheory
-  P : kernel Θ 𝒳
+structure estimationProblem (Θ 𝒴 𝒵 : Type*) [mΘ : MeasurableSpace Θ]
+    [m𝒴 : MeasurableSpace 𝒴] [m𝒵 : MeasurableSpace 𝒵] :=
   /-- The objective function. -/
   y : Θ → 𝒴
   y_meas : Measurable y
   /-- The cost (or loss) function. -/
   ℓ : 𝒴 × 𝒵 → ℝ≥0∞
   ℓ_meas : Measurable ℓ
 
-/-- Given an estimation problem `E` and a kernel `η : 𝒳 ⇝ 𝒳'`, we can consider a new estimation
-problem where the kernel is given by the composition between `η` and `E.P`. -/
-@[simps]
-noncomputable
-def estimationProblem.comp (E : estimationProblem Θ 𝒳 𝒴 𝒵) (η : kernel 𝒳 𝒳') :
-    estimationProblem Θ 𝒳' 𝒴 𝒵 where
-  P := η ∘ₖ E.P
-  y := E.y
-  y_meas := E.y_meas
-  ℓ := E.ℓ
-  ℓ_meas := E.ℓ_meas
-
 /-- Given an estimation problem `E` and a measurable function `f : Θ' → Θ`, we can consider a new
 estimation problem where the kernel is given by the pullback of `E.P` through `f`. -/
 @[simps]
 noncomputable
-def estimationProblem.comap (E : estimationProblem Θ 𝒳 𝒴 𝒵) (f : Θ' → Θ) (hf : Measurable f) :
-    estimationProblem Θ' 𝒳 𝒴 𝒵 where
-  P := kernel.comap E.P f hf
+def estimationProblem.comap (E : estimationProblem Θ 𝒴 𝒵) (f : Θ' → Θ) (hf : Measurable f) :
+    estimationProblem Θ' 𝒴 𝒵 where
   y := E.y ∘ f
   y_meas := E.y_meas.comp hf
   ℓ := E.ℓ
   ℓ_meas := E.ℓ_meas
 
-@[simp]
-lemma estimationProblem.comp_comp (E : estimationProblem Θ 𝒳 𝒴 𝒵) (κ : kernel 𝒳 𝒳')
-    (η : kernel 𝒳' 𝒳'') [IsSFiniteKernel η] :
-    (E.comp κ).comp η = E.comp (η ∘ₖ κ) := by
-  ext <;> simp [kernel.comp_assoc]
-
-/-- Given an estimation problem `E` and a kernel `κ : (Θ × 𝒳) ⇝ 𝒳'`, we can consider a new
-estimation problem where the kernel is given by the composition product of `E.P` and `κ`. -/
-@[simps]
-noncomputable
-def estimationProblem.compProd (E : estimationProblem Θ 𝒳 𝒴 𝒵) (κ : kernel (Θ × 𝒳) 𝒳') :
-    estimationProblem Θ (𝒳 × 𝒳') 𝒴 𝒵 where
-  P := E.P ⊗ₖ κ
-  y := E.y
-  y_meas := E.y_meas
-  ℓ := E.ℓ
-  ℓ_meas := E.ℓ_meas
-
 end EstimationProblem
 
-/-- The risk of an estimator `κ` on an estimation problem `E` at the parameter `θ`. -/
+variable {E : estimationProblem Θ 𝒴 𝒵}
+
+/-- The risk of an estimator `κ` on an estimation problem `E` with data generating kernel `P`
+at the parameter `θ`. -/
 noncomputable
-def risk (E : estimationProblem Θ 𝒳 𝒴 𝒵) (κ : kernel 𝒳 𝒵) (θ : Θ) : ℝ≥0∞ :=
-  ∫⁻ z, E.ℓ (E.y θ, z) ∂((κ ∘ₖ E.P) θ)
+def risk (E : estimationProblem Θ 𝒴 𝒵) (P : kernel Θ 𝒳) (κ : kernel 𝒳 𝒵) (θ : Θ) : ℝ≥0∞ :=
+  ∫⁻ z, E.ℓ (E.y θ, z) ∂((κ ∘ₖ P) θ)
 
-/-- The bayesian risk of an estimator `κ` on an estimation problem `E` with respect to
-a prior `π`. -/
+@[simp]
+lemma risk_comp_assoc (E : estimationProblem Θ 𝒴 𝒵) (P : kernel Θ 𝒳) [IsSFiniteKernel P]
+    (κ : kernel 𝒳'' 𝒵) (θ : Θ)
+    (η : kernel 𝒳 𝒳') [IsMarkovKernel η] (η' : kernel 𝒳' 𝒳'') [IsMarkovKernel η'] :
+    risk E (η' ∘ₖ (η ∘ₖ P)) κ θ = risk E (η' ∘ₖ η ∘ₖ P) κ θ := by
+  rw [risk, risk]
+  congr 1
+  simp_rw [kernel.comp_apply]
+  congr 1
+  rw [Measure.comp_assoc]
```

Comment:
```suggestion
  rw [kernel.comp_assoc]
```

Reply:
Do we actually need this lemma?
It is just an application of `kernel.comp_assoc`.

Reply:
Somehow I forgot about kernel.comp_assoc ? I don't know why.